sui-execution/latest/sui-adapter/src/execution_engine.rs

```diff
@@ -362,7 +330,92 @@ fn execute_transaction<Mode: ExecutionMode>(
     } // else, we're in the genesis transaction which mints the SUI supply, and hence does not satisfy SUI conservation, or
       // we're in the non-production dev inspect mode which allows us to violate conservation
       // === end SUI conservation checks ===
-    (cost_summary, result)
+    result
+}
+
+#[instrument(name = "check_meter_limit", level = "debug", skip_all)]
+fn check_meter_limit(
+    temporary_store: &mut TemporaryStore<'_>,
+    gas_charger: &mut GasCharger,
+    protocol_config: &ProtocolConfig,
+    metrics: Arc<LimitsMetrics>,
+) -> Result<(), ExecutionError> {
+    let effects_estimated_size = temporary_store.estimate_effects_size_upperbound();
+
+    // Check if a limit threshold was crossed.
+    // For metered transactions, there is not soft limit.
+    // For system transactions, we allow a soft limit with alerting, and a hard limit where we terminate
+    match check_limit_by_meter!(
+        !gas_charger.is_unmetered(),
+        effects_estimated_size,
+        protocol_config.max_serialized_tx_effects_size_bytes(),
+        protocol_config.max_serialized_tx_effects_size_bytes_system_tx(),
+        metrics.excessive_estimated_effects_size
+    ) {
+        LimitThresholdCrossed::None => (),
+        LimitThresholdCrossed::Soft(_, limit) => {
+            warn!(
+                effects_estimated_size = effects_estimated_size,
+                soft_limit = limit,
+                "Estimated transaction effects size crossed soft limit",
+            )
+        }
+        LimitThresholdCrossed::Hard(_, lim) => {
+            return Err(ExecutionError::new_with_source(
+                ExecutionErrorKind::EffectsTooLarge {
+                    current_size: effects_estimated_size as u64,
+                    max_size: lim as u64,
+                },
+                "Transaction effects are too large",
+            ))
+        }
+    };
+
+    Ok(())
+}
+
+#[instrument(name = "check_written_objects_limit", level = "debug", skip_all)]
+fn check_written_objects_limit<Mode: ExecutionMode>(
+    temporary_store: &mut TemporaryStore<'_>,
+    gas_charger: &mut GasCharger,
+    protocol_config: &ProtocolConfig,
+    metrics: Arc<LimitsMetrics>,
+) -> Result<(), ExecutionError> {
+    // This limit is only present in Version 3 and up, so use this to gate it
+    if let (Some(normal_lim), Some(system_lim)) = (
+        protocol_config.max_size_written_objects_as_option(),
+        protocol_config.max_size_written_objects_system_tx_as_option(),
+    ) {
```

Comment:
calling on @oxade to confirm but actually mainnet started at protocol version 4 and there is no previous version no more, so this check could actually be removed and the values fetched directly.
I think it may be a good cleanup

Reply:
removed